lptim: stm32l4: support of the lptimer

FRASTM opened the pull request:
This patch enables the LP Timer driver for the stm32l4r5 soc, to be used in tickless kernel. The LPTimer is clocked by the LSI. It provides the following APIs:
z_clock_set_timeout
z_clock_elapsed
z_clock_driver_init
and irq handler

Signed-off-by: Francois Ramu francois.ramu@st.com

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+	/* maximise to TIMEBASE */
+	if (next_arr > LPTIM_TIMEBASE) {
+		next_arr = LPTIM_TIMEBASE;
+	}
```
Will counter automatically restart from 0 ? 
Otherwise in case next_arr < cnt_value, you need to also program to LPTIM_TIMEBASE ?

FRASTM replied:
yes, LPTIM is reset when ARR and CNT register values matches

LMESTM reviewed (commented):
Few comments and questions in PR but this overall looks promising !

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+#define COUNT_PER_MS (LPTIM_CLOCK / (LPTIM_PRESCALER * 1000))
+
+/* assuming one TICKS is 10 ms long */
+#define MS_PER_TICK 10
```
even with #ifdef CONFIG_SYS_CLOCK_EXISTS
this constant gives a warning: division by zero [-Wdiv-by-zero]  when using MS_PER_TICK

FRASTM replied:
constant is based on the CONFIG_SYS_CLOCK_TICKS_PER_SEC

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+#ifdef CONFIG_TICKLESS_KERNEL
+
+	if (ticks == K_FOREVER) {
+		/* disable LPTIM */
```
with K_FOREVER, as the timeout delay is infinite, only a re-init is possible

LMESTM replied:
So is clear that Kernel will call again the init ? 
Isn't it possible that system is waken-up by an external trigger and therefore the kernel just makes a new call to z_clock_set_timeout ?

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+	/*LL_LPTIM_EnableResetAfterRead(LPTIM1);*/
+	LL_LPTIM_TrigSw(LPTIM1);
+	LL_LPTIM_SetInput1Src(LPTIM1, LL_LPTIM_INPUT1_SRC_GPIO);
+	LL_LPTIM_SetInput2Src(LPTIM1, LL_LPTIM_INPUT2_SRC_GPIO);
```
ok

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+	LL_LPTIM_DisableTimeout(LPTIM1);
+	/*LL_LPTIM_EnableResetAfterRead(LPTIM1);*/
+	LL_LPTIM_TrigSw(LPTIM1);
+	LL_LPTIM_SetInput1Src(LPTIM1, LL_LPTIM_INPUT1_SRC_GPIO);
```
ok

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+	/* configure the LPTIM1 counter */
+	LL_LPTIM_SetClockSource(LPTIM1, LL_LPTIM_CLK_SOURCE_INTERNAL);
+	/* configure the LPTIM1 prescaler is 2 */
+	LL_LPTIM_SetPrescaler(LPTIM1, LL_LPTIM_PRESCALER_DIV2);
```
actually LL_LPTIM_PRESCALER_DIV2 is not 0x02 but 0x200

FRASTM replied:
and LL_LPTIM_PRESCALER_DIV8 is 0x600

erwango commented on drivers/timer/Kconfig:
```diff
@@ -165,6 +165,19 @@ config NRF_RTC_TIMER
 
+config STM32_LPTIM_TIMER
+	bool "STM32 Low Power Timer"
+	depends on SOC_FAMILY_STM32
```
Lets limit to SOC_SERIES_STM32L4X for now

FRASTM replied:
CONFIG_STM32_LPTIM_TIMER is defined in the application (sample/subsys/power/stm32_pm/) to avoid disturbing other examples with lptimer

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,270 @@
+	LL_RCC_ForceBackupDomainReset();
+	LL_RCC_ReleaseBackupDomainReset();
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) &&	\
+	!defined(CONFIG_SOC_SERIES_STM32F2X)
```
Remove references to series other than L4 for now

FRASTM replied:
removed

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,270 @@
+#endif
+
+	/* Clear the event flag and possible pending interrupt */
+	IRQ_CONNECT(LPTIM1_IRQn, 1, lptim_irq_handler, 0, 0);
```
Please use data from device tree.

FRASTM replied:
LPTIM1_IRQn is DT_LPTIM_1_IRQ defined by the DT_ST_STM32_TIMERS_40007C00_IRQ_0 in the dts

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+#error "LPTIM_CLOCK not valid"
+#endif
+#define LPTIM_MAX_COUNT 0xFFFF
+#define LPTIM_PRESCALER 2
```
Is that still required?

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,270 @@
+#define LPTIM_CLOCK LSI_VALUE
+
+/* LPTIM counter for a 4s timebase */
+#if (LPTIM_CLOCK == LSI_VALUE)
```
Please use Kconfig for configuration

FRASTM replied:
get the  LPTIM_CLOCK value through KConfig

erwango commented on soc/arm/st_stm32/stm32l4/Kconfig.series:
```diff
@@ -12,6 +12,7 @@ config SOC_SERIES_STM32L4X
 	select CPU_HAS_FPU
 	select SOC_FAMILY_STM32
 	select HAS_STM32CUBE
+	select STM32_LPTIM_TIMER
```
Shouldn't it be conditional to some other pm feature ?

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,279 @@
+
+/*
+ * This is just a getting started point.
+ *
```
Maybe we can also elaborate here about the choice of timebase and presecaler.

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi_defconfig:
```diff
@@ -24,6 +24,9 @@ CONFIG_PINMUX=y
 CONFIG_GPIO=y
 
+# Enable LPTIM
+CONFIG_STM32_LPTIM_TIMER=y
```
I'd prefer we keep LPTIM opionnal for now.
Could be enabled in a dedicated sample though

erwango commented on samples/boards/stm32l4r5/lptim/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
Please move to `samples/subsys/power/stm32`

erwango commented on samples/boards/stm32l4r5/lptim/prj.conf:
```diff
@@ -0,0 +1,16 @@
+CONFIG_BT=n
```
Should not be needed

erwango commented on samples/boards/stm32l4r5/lptim/prj.conf:
```diff
@@ -0,0 +1,16 @@
+CONFIG_SYS_PM_STATE_LOCK=y
+#CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1=5000
+#CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_2=15000
+CONFIG_GPIO=y
```
Should not be needed as GPIO is default y on all boards

erwango commented on samples/boards/stm32l4r5/lptim/prj.conf:
```diff
@@ -0,0 +1,16 @@
+CONFIG_TICKLESS_KERNEL=y
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=100
+CONFIG_SYS_PM_STATE_LOCK=y
+#CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1=5000
```
Please remove these

erwango commented on samples/boards/stm32l4r5/lptim/prj.conf:
```diff
@@ -0,0 +1,16 @@
+CONFIG_BT=n
+CONFIG_TIMESLICING=n
+CONFIG_STM32_LPTIM_TIMER=y
+CONFIG_SYS_POWER_MANAGEMENT=n
```
Mayeb some of the symbols set to `n` could be cleaned up.

erwango commented on samples/boards/stm32l4r5/lptim/prj.conf:
```diff
@@ -0,0 +1,16 @@
+#CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1=5000
+#CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_2=15000
+CONFIG_GPIO=y
+CONFIG_EXTI_STM32=y
```
This is default `y` on all STM32 so not requested neither

erwango commented on samples/boards/stm32l4r5/lptim/sample.yaml:
```diff
@@ -0,0 +1,12 @@
+sample:
+  name: Low Power State Sample for stm32l4r5
```
Please use full board name instead of `stm32l4r5`

erwango commented on samples/boards/stm32l4r5/lptim/src/sample.h:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
```
Please add your copyright.

erwango reviewed (changes requested):
Second round of comments.
Please take car of first's commit title as well.

FRASTM commented:
add samples/subsys/power/stm32_pm/lptim example to use the lptimer in the tickless kernel
with a CONFIG_STM32_LPTIM_TIMER

FRASTM commented on drivers/timer/Kconfig:
```diff
@@ -165,6 +165,23 @@ config NRF_RTC_TIMER
+	bool "STM32 Low Power Timer 1"
+	depends on LPTIM_TIMER
+
+config LPTIM_2
```
STM32_LPTIM_2 removed

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+		z_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL)
+			? dticks : 1);
+	}
+#endif
```
done

FRASTM commented:
use Kconfig for configuration of the LPTIM_CLOCK : either LSI_VALU or LSE_VALUE through 
one of LPTIM_CLOCK_LSI or LPTIM_CLOCK_LSE

FRASTM commented on drivers/timer/Kconfig:
```diff
@@ -165,6 +165,23 @@ config NRF_RTC_TIMER
 	  Counter NRF_RTC1 and provides the standard "system clock driver"
 	  interfaces.
 
+config LPTIM_TIMER
```
changed to STM32_LPTIM_TIMER in Kconfig

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+			LPTIM_PRESCALER * 1000) / LPTIM_CLOCK)) + 1;
+		z_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL)
+			? dticks : 1);
+	}
```
u32_t dticks = (cnt_value * CONFIG_SYS_CLOCK_TICKS_PER_SEC)
				/ LPTIM_CLOCK;

erwango commented on drivers/timer/Kconfig:
```diff
@@ -165,6 +165,25 @@ config NRF_RTC_TIMER
+	  This module implements a kernel device driver for the LowPower Timer
+	  and provides the standard "system clock driver" interfaces.
+
+config LPTIM_CLOCK_LSI
```
Since it's one or the other, use a choice semantic

FRASTM replied:
with choice on STM32_LPTIM_CLOCK
STM32_LPTIM_CLOCK_LSI or STM32_LPTIM_CLOCK_LSE in the Kconfig - timer driver configuration options

erwango commented on drivers/timer/Kconfig:
```diff
@@ -165,6 +165,25 @@ config NRF_RTC_TIMER
+	  This module implements a kernel device driver for the LowPower Timer
+	  and provides the standard "system clock driver" interfaces.
+
+config LPTIM_CLOCK_LSI
```
add STM32 namespace, for instance STM32_LPTIM_CLOCK_LSI

FRASTM replied:
defining STM32_LPTIM_CLOCK

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,276 @@
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
+	LL_APB1_GRP1_ReleaseReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+	/* select LSI as LPTIM1 clock source */
+#if (LPTIM_CLOCK == LSI_VALUE)
```
Use dedicated Kconfig symbol

FRASTM replied:
with choice on STM32_LPTIM_CLOCK 
STM32_LPTIM_CLOCK_LSI or STM32_LPTIM_CLOCK_LSE

erwango commented on dts/arm/st/l4/stm32l4r5.dtsi:
```diff
@@ -145,6 +145,7 @@
 			status = "disabled";
 			label= "OTGFS";
 		};
+
```
Remove this line

FRASTM replied:
ok

erwango commented on soc/arm/st_stm32/stm32l4/Kconfig.series:
```diff
@@ -13,6 +13,8 @@ config SOC_SERIES_STM32L4X
 	select HAS_STM32CUBE
 	select CPU_HAS_ARM_MPU
+	select LPTIM_CLOCK_LSI
+	select LPTIM_CLOCK_LSE
```
Can't be both...
Btw, is that required?

FRASTM replied:
removed

FRASTM commented:
add the lptim structure in the DTS as a timer

FRASTM commented:
move the STM32_LPTIM_CLOCK configuration to the drivers/timer/KConfig

FRASTM commented:
add a specific drivers/timer/Kconfig for stm32 : `drivers/timer/Kconfig.stm32_lptim`

FRASTM commented:
using lptimer to set timeout with example application using the k_sleep()
In a second step, the low power modes ( power: stm32l4r5: support of the Low Power Mode #19026 )
will be merged.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,271 @@
+/* nb of LPTIM counter unit per kernel tick  */
+#define COUNT_PER_TICK (LPTIM_CLOCK / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+struct device;	       /* forward declaration; type is not used. */
```
Is that still required ?

erwango replied:
Please check this.

erwango reviewed (changes requested):
Minor comments.
Last, I don't go in code details, but drivers is stricylty made to use LPTIM1 while PR provides LPTIM2 definition as well. So either we clean up PR to provide only LPTIM1 references, either driver is adapated ot use LPTIM2 as well.
In both cases, I think we'd benefit to move driver to use `LPTIM_TypeDef *lptim` structs as of now instead of hardcoded CMSIS defined `LPTIM1`.

FRASTM commented:
give more precision to k_sleep :  timing verified from 7ms to 20s

FRASTM commented:
According to the ref man, writing to the AutoReload register (ARR) is valid when the ARROK bit is set in the ISR register --> poll ARROK bit after writing the ARR.

FRASTM commented:
add more precision to timing, in case the LPTIM is clocked by LSE

FRASTM commented:
new sample/subsys/power/stm32_pm/lptim application

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,279 @@
+	/* enable LPTIM clock source */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
+	LL_APB1_GRP1_ReleaseReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+	/* select LSI as LPTIM1 clock source */
```
This comment could be removed I think.

FRASTM replied:
yes

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,279 @@
+	LL_RCC_SetLPTIMClockSource(LL_RCC_LPTIM1_CLKSOURCE_LSI);
+
+#else /* CONFIG_STM32_LPTIM_CLOCK_LSI */
+#if defined(LL_APB1_GRP1_PERIPH_PWR)
```
Any case where this is not defined? (on L4 for now)

erwango replied:
Please check this.

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,279 @@
+	/*LL_LPTIM_EnableResetAfterRead(LPTIM1);*/
+	LL_LPTIM_TrigSw(LPTIM1);
+#ifdef LL_LPTIM_INPUT1_SRC_GPIO
+	LL_LPTIM_SetInput1Src(LPTIM1, LL_LPTIM_INPUT1_SRC_GPIO);
```
Can you elaborate why is this line required actually?

FRASTM replied:
this is the default input pin configuration

erwango replied:
So maybe this is not needed. We don' t use inputs anyway currently, so maybe this is clearer if we remove these.
.

erwango commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -300,6 +300,30 @@
+			label = "LPTIM_1";
+		};
+
+		lptim2: timers@40009400 {
```
Remove lptim2 if we don't have code to use it.

FRASTM replied:
ok

erwango commented on soc/arm/st_stm32/stm32l4/dts_fixup.h:
```diff
@@ -323,4 +323,18 @@
+#define DT_LPTIM_1_CLOCK_BITS	DT_ST_STM32_TIMERS_40007C00_CLOCK_BITS_0
+#define DT_LPTIM_1_CLOCK_BUS	DT_ST_STM32_TIMERS_40007C00_CLOCK_BUS_0
+
+#define DT_LPTIM_2_BASE_ADDRESS	DT_ST_STM32_LPTIM_40009400_BASE_ADDRESS
```
Remove LPTIM2 if no code to use it

FRASTM replied:
ok

erwango commented on samples/subsys/power/stm32_pm/lptim/src/main.c:
```diff
@@ -0,0 +1,48 @@
+	gpio_port = device_get_binding(PORT);
+
+	/* Configure Button 1 as deep sleep trigger event (no Pull up on PA0) */
+	gpio_pin_configure(gpio_port, SW, GPIO_DIR_IN);
```
Seems SW code is not used for now.

FRASTM replied:
changed sample to increase timeout when button is pressed

FRASTM commented on samples/subsys/power/stm32_pm/lptim/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_TIMESLICING=n
+CONFIG_STM32_LPTIM_TIMER=y
+CONFIG_CORTEX_M_SYSTICK=n
```
Kconfig.defconfig.series has 
if STM32_LPTIM_TIMER
config CORTEX_M_SYSTICK
	default n
endif # STM32_LPTIM_TIMER

but sample prj.conf still requires
CONFIG_CORTEX_M_SYSTICK=n

FRASTM commented:
considering the comments
and changing the sample demo to increase timeout when button is pressed

FRASTM commented:
re-push to pass 'Shippable'

erwango commented on samples/subsys/power/stm32_pm/lptim/src/main.c:
```diff
@@ -0,0 +1,66 @@
+
+/* 1000 msec = 1 sec */
+
+#define WAIT_TIME 100
```
Let's set to 500ms to make it more the sample result more visible

FRASTM replied:
changed

erwango commented on samples/subsys/power/stm32_pm/lptim/src/main.c:
```diff
@@ -0,0 +1,66 @@
+		/* Set pin to HIGH/LOW */
+		gpio_pin_write(gpio_port0, LED0, LED_ON);
+
+		k_sleep(timeout);
```
Let's add use of `k_uptime` and `k_uptime_delta` to see if we manage to remain coherent

FRASTM replied:
done in main.c function button_pressed

erwango commented on samples/subsys/power/stm32_pm/lptim/sample.yaml:
```diff
@@ -0,0 +1,12 @@
+sample:
```
Please add a README file

FRASTM replied:
through a "DEMO_DESCRIPTION" in the main.c

erwango commented on soc/arm/st_stm32/stm32l4/Kconfig.defconfig.series:
```diff
@@ -25,6 +24,13 @@ config GPIO_STM32_PORTH
 
+if STM32_LPTIM_TIMER
+
+config CORTEX_M_SYSTICK
```
Remove this.
Let's do configuration in applications for now

FRASTM commented:
resubmit

FRASTM commented:
changed drivers/timer/Kconfig.stm32_lptim

FRASTM commented:
move CONFIG_CORTEX_M_SYSTICK to sample and not in soc KCONFIG

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -8,7 +8,7 @@
 
 if SOC_FAMILY_STM32
-
+ 
```
File should be left untouched

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -26,7 +26,6 @@ config UART_STM32
 	default y
 
 endif #SERIAL
-
```
File should be left untouched

erwango commented on dts/arm/st/wb/stm32wb.dtsi:
```diff
@@ -272,11 +272,24 @@
 			#io-channel-cells = <1>;
 		};
 
+
```
Remove this line

erwango commented on samples/subsys/power/stm32_pm/lptim/sample.yaml:
```diff
@@ -0,0 +1,12 @@
+sample:
+  name: Low Power Timer Sample for nucleo_l4r5zi
```
Update sample name

erwango commented on drivers/timer/CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ zephyr_sources_if_kconfig(              loapic_timer.c)
 zephyr_sources_if_kconfig(              apic_timer.c)
 zephyr_sources_ifdef(CONFIG_ALTERA_AVALON_TIMER altera_avalon_timer_hal.c)
 zephyr_sources_if_kconfig(              nrf_rtc_timer.c)
+zephyr_sources_if_kconfig(              stm32_lptim_timer.c)
```
Update commit title, should be something like:
`drivers/timer: stm32: Add ....`

erwango commented on drivers/timer/Kconfig.stm32_lptim:
```diff
@@ -0,0 +1,44 @@
+
+config STM32_LPTIM_TIMER
+	bool "STM32 Low Power Timer"
+	depends on (SOC_SERIES_STM32L4X)
```
Should it be `(... || SOC_SERIES_STM32WBX)`?

skakri replied:
I'm not sure about the scope of these changes (L4x only?) but similarly `|| SOC_SERIES_STM32L0X` could be added as well. `LPTIM` is supported by almost any STM32L series MCUs with some exceptions: all of `STM32L1x` range and these six MCUs: `STM32L051C8Ux`, `STM32L052C8Ux`, `STM32L053C8Ux`, `STM32L063C8Ux`, `STM32L073CZUx`, `STM32L552QCUx`.

FRASTM commented:
fix common/Kconfig.defconfig.series

FRASTM commented:
take L4 serie only
add commit for test/kernel/sleep

FRASTM commented:
manage CONFIG_STM32_LPTIM_TIMER directly in the soc
and CONFIG_STM32_LPTIM_CLOCK_LSI in the board
Consequently proj.conf could change to CONFIG_STM32_LPTIM_CLOCK_LSE in the sample application
config CORTEX_M_SYSTICK and config TICKLESS_KERNEL are defined in the common defconfig

FRASTM commented:
CONFIG_TICKLESS_KERNEL not required,
TICKLESS_CAPABLE is already configured in the driver/timer/Kconfig.stm32_lptim

FRASTM commented:
rebase

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi_defconfig:
```diff
@@ -14,6 +14,9 @@ CONFIG_BOARD_NUCLEO_L4R5ZI=y
 # 80MHz system clock in 'normal power' mode
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=80000000
 
+# clock for the LPTIMER
```
Please put this in board Kconfig.defconfig under if lp timer condition

FRASTM replied:
we could just have in nucleo_xx_defconfig of the board:
 # enable lptimer and its clock source
CONFIG_STM32_LPTIM_TIMER=y
CONFIG_STM32_LPTIM_CLOCK_LSE=y

FRASTM replied:
choice for LSE or LSI clock for LPTIMER clock is now in the Kconfig.defconfig for the board, under STM32_LPTIM_TIMER condition:
`if STM32_LPTIM_TIMER

choice STM32_LPTIM_CLOCK
	default STM32_LPTIM_CLOCK_LSI

endchoice

endif #STM32_LPTIM_TIMER
`

FRASTM replied:
the application must now define the CONFIG_STM32_LPTIM_TIMER to enable the LPTIMER feature in **prj.conf**
> 
CONFIG_STM32_LPTIM_TIMER=y

erwango reviewed (changes requested):
This is close

FRASTM commented on soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx:
```diff
@@ -15,6 +15,9 @@ config NUM_IRQS
 	int
 	default 95
 
+config STM32_LPTIM_TIMER
```
ok

FRASTM committed:
boards: arm: st_stm32: add lptimer to nucleo_l4r5zi boards

This patch introduces the support of the LowPower Timer
     for the STM32L4xx from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
change config to application proj.conf, no more to board config

FRASTM commented:
It could happen that test/kernel/timer/timer_monotic fails :
    "diff = -80000 (t_last = 113200000 : t_now = 113120000);i = 972"

Test fails : tests/kernel/fifo/fifo_timeout : test_timeout_threads_pend_fail_on_fifo 
Test fails : tests/kernel/lifo/lifo_usage : test_timeout_threads_pend_on_lifo

FRASTM commented:
rebase

FRASTM commented:
rebase

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+#define COUNT_PER_TICK (LPTIM_CLOCK / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+static u32_t last_cnt;
+static u32_t accumulated_lptim_ms;
```
Seems to be equal to 49 days -- quite okay but wrap-around will happen sooner that later in a real product ... moving to u64_t would make it less likely to happen

FRASTM replied:
we changed to store the total nb of LPTIM count unit, which is shorter.

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+	ARG_UNUSED(unused);
+
+	if ((LL_LPTIM_IsActiveFlag_ARRM(LPTIM1) != 0)
+		&& (LL_LPTIM_IsEnabledIT_ARRM)(LPTIM1) != 0) {
```
isnt there an extra parenthesis that should be removed ?
(LL_LPTIM_IsEnabledIT_ARRM)(LPTIM1)
vs.
(LL_LPTIM_IsEnabledIT_ARRM(LPTIM1)

FRASTM replied:
ok

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+
+		k_spinlock_key_t key = k_spin_lock(&lock);
+
+		/* LPTIM1 CNT could have already been reset */
```
Comment is not so clear, what are we storing in cnt_value ?
Could CNT be reset already, or has it for sure been reset already ?

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+
+		/* a next timeout will be raised */
+		LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
+		/* reset last_cnt for a next call to 'elapsed' */
```
what does the comment about last_cnt refer to ?

FRASTM replied:
unused

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+		/* increase the total nb of lptim count
+		 * used in the z_timer_cycle_get_32() function
+		 */
+		accumulated_lptim_ms += ((cnt_value * 1000) / LPTIM_CLOCK);
```
Are we confident about the precision ?
couldn't we store accumulated number of counts and only convert to ms when needed (or in a MACRO) ?

FRASTM replied:
yes : we changed to store nb of LPTIM counter unit (instead of ms).
converting to ms gives unaccuracy which is "announced" and considered by next the kernel call to 'set timeout()'

FRASTM replied:
in anycase the z_timer_cycle_get_32() function expects an u32 value

FRASTM replied:
var is now named accumulated_lptim_cnt

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+		 */
+		accumulated_lptim_ms += ((cnt_value * 1000) / LPTIM_CLOCK);
+
+		/* announce the elapsed time in ms, cnt_value is 16bit */
```
cnt_value is declared as u32_t ... not like the comment says ..

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+		LL_LPTIM_Enable(LPTIM1);
+
+		/* Set the Autoreload value once the timer is enabled */
+		LL_LPTIM_SetAutoReload(LPTIM1, COUNT_PER_TICK);
```
The 2 different initialization could be worth a few comments.
so in case of TICKS mode, the auto reload is set to COUNT_PER_TICK so there will be an interrupt for each tick and the counter will always reset right ?  and the comparator will always be kept to 0 and never modified ?

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+		/* Set the Autoreload value once the timer is enabled */
+		LL_LPTIM_SetAutoReload(LPTIM1, LPTIM_TIMEBASE);
+
+		/* do not use compare yet, set the compare to period */
```
The 2 different initialization could be worth a few comments.
in case of tickless the comparator will be later modified depending on the Kernel request ?

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+
+	/* minimise (ie 10 counter cycles) to trig timer */
+	if ((next_arr - LL_LPTIM_GetAutoReload(LPTIM1)) < 10) {
+		next_arr = LL_LPTIM_GetAutoReload(LPTIM1) + 10;
```
Not clear what is being done here ... Isn't Auto reload equal to LPTIM_TIMEBASE ?
so LPTIM_TIMEBASE + 10 will be more than LPTIM_TIMEBASE  ..
and next check below will make it back to LPTIM_TIMEBASE and possibly less than 10 units away ...
How to make sure it will fire in this case ?

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+	/* run timer and wait for the reload match */
+	LL_LPTIM_SetAutoReload(LPTIM1, next_arr);
+	/* poll ARROK bit to validate write operation to ARR register */
+	while (LL_LPTIM_IsActiveFlag_ARROK(LPTIM1) == 0) {
```
This loop can induce a very latency, by doing so you will always loop and "waste" time here
maybe it would be better to add this check at the beginning of the function to verify that the previous programmin was okay and therefore to wait only in case the previous programming has not completed yet ..  ?

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+	/* gives the value of LPTIM1 counter (ms)
+	 * since the previous 'announce'
+	 */
+	u32_t ret = ((cnt_value - last_cnt) * 1000) / LPTIM_CLOCK;
```
last_cnt can be modified out of this function so maybe this line needs to be protected by k_spin_unlock(&lock, key);  ?

LMESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,285 @@
+	k_spin_unlock(&lock, key);
+
+	/* give accumulated time in ms */
+	u32_t ret = accumulated_lptim_ms + (cnt_value * 1000) / LPTIM_CLOCK;
```
same ehre: accumulated_lptim_ms can be modified in interrupts handler, so need to protect this line with spin lock ?

FRASTM replied:
ok

FRASTM commented:
change after review

FRASTM commented:
rebase and fix conflict

FRASTM commented:
checkpath

FRASTM commented:
rebase

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,292 @@
+	 */
+	volatile u32_t cnt_value = LL_LPTIM_GetCounter(LPTIM1);
+	while (last_cnt < cnt_value) {
+		last_cnt = LL_LPTIM_GetCounter(LPTIM1);
```
no
 this is the way the ref manual is requesting to read the LPTIM counter:
" It should be noted that to read reliably the content of the LPTIM_CNT register, two successive read accesses must be performed and compared"

LMESTM replied:
But you need the comparison to take care of case like I describe above. In the way it is now, there are rare cases where it would get stuck in the while loop for 2 seconds ... 
Here is an alternative proposal from mbed-os for consecutive read and while loop which will not get stuck for a complete wrap-around: 
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/lp_ticker.c#L348

FRASTM replied:
reading the LPTIM counter register is now:
`	while (lp_time != LL_LPTIM_GetCounter(LPTIM1)) {
		lp_time = LL_LPTIM_GetCounter(LPTIM1);
	}
`

FRASTM committed:
soc: arm: st_stm32: add driver lptimer management to stm32l4 series

This patch introduces the support of the Low Power Timer
     for the STM32L4xx from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,273 @@
+	LL_PWR_EnableBkUpAccess();
+	LL_RCC_ForceBackupDomainReset();
+	LL_RCC_ReleaseBackupDomainReset();
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) &&	\
```
flagged with LL_APB1_GRP1_PERIPH_PWR

FRASTM commented:
fix merge error in dts for L4

FRASTM committed:
driver: timer: st_stm32: add lptimer management to stm32xx series

This patch introduces the support of the LowPower Timer
 for the STM32xx from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
changing the loop for reading LPTIM counter
adjust timing accuracy

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -0,0 +1,306 @@
+	LL_LPTIM_ClearFlag_ARROK(LPTIM1);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+	if (ticks == K_FOREVER) {
```
> So is clear that Kernel will call again the init ?
Isn't it possible that system is waken-up by an external trigger and therefore the kernel just makes a new call to z_clock_set_timeout ?

Can you respond question from @LMESTM ?

FRASTM replied:
as specified by the system_timer.h for the Set system clock timeout
K_FOREVER is disabling the timer interrupts, so that in our case we reset the LPTIM timer
>  * Note that ticks can also be passed the special value K_FOREVER,
>  * indicating that no future timer interrupts are expected or required
>  * and that the system is permitted to enter an indefinite sleep even
>  * if this could cause rollover of the internal counter

erwango commented:
@galak

galak commented on samples/subsys/power/stm32_pm/lptim/sample.yaml:
```diff
@@ -0,0 +1,14 @@
+  name: Low Power Timer Sample for nucleo board
+tests:
+  sample.power.pm.stm32_lptim.lsi:
+    platform_whitelist: nucleo_l4r5zi
```
Any reason this is limited to just nucleo_l4r5zi?

FRASTM replied:
This is the board I have tested on
I am adding the disco_l475_iot1

FRASTM committed:
boards: arm: st_stm32: add lptimer to disco_l475_iot1 board

This patch introduces the support of the LowPower Timer
     for the STM32L4xx from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
- adding the disco_l475_iot1

- removing the sample application
